core/client/src/main/java/com/thousandeyes/sdk/pagination/Paginator.java

```diff
@@ -0,0 +1,101 @@
+package com.thousandeyes.sdk.pagination;
+
+import java.lang.reflect.InvocationTargetException;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.function.Function;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import com.thousandeyes.sdk.client.ApiException;
+
+
+
+public class Paginator<T, R> implements Iterable<T> {
+    private static final Pattern CURSOR_PATTERN = Pattern.compile("cursor=([^&]*)");
+    private final PaginatedApiCall<R> apiCall;
+    private final Function<R, List<T>> dataExtractor;
+
+    public Paginator(PaginatedApiCall<R> apiCall, Function<R, List<T>> dataExtractor)
+    {
+        this.apiCall = apiCall;
+        this.dataExtractor = dataExtractor;
+    }
+
+    private String extractCursor(String href) {
+        if (href == null) {
+            return null;
+        }
+
+        var matcher = CURSOR_PATTERN.matcher(href);
+        if (matcher.find()) {
+            return URLDecoder.decode(matcher.group(1), StandardCharsets.UTF_8);
+        }
+
+        return null;
+    }
+
+    @Override public Iterator<T> iterator() {
+        return new PaginatorIterator();
+    }
+
+    public Stream<T> stream() {
+        return StreamSupport.stream(spliterator(), false);
+    }
+
+    private class PaginatorIterator implements Iterator<T> {
+        private String cursor = null;
+        private Iterator<T> currentBatchIterator = null;
+        private boolean hasNextBatch = true;
+
+        @Override
+        public boolean hasNext() {
+            if (currentBatchIterator == null || !currentBatchIterator.hasNext()) {
+                if (hasNextBatch) {
+                    fetchNextBatch();
+                }
+            }
+            return currentBatchIterator != null && currentBatchIterator.hasNext();
+        }
```

Comment:
imo the logic to fetch the initial batch should be in the `next` method instead of the `hasNext`.
Haven't tested but i think changing the `hasNext` method to just return the `hasNextBatch` property 
and changing the `next` method to call the `fetchNextBatch` before returning the `currentBatchIterator`, should do it 🤔

Reply:
This was not possible because the `audit-user-events` api returns a next link then returns an empty array, so we need to know whether there are elements on the next pager prior to the `next()` method call.